### ISSUE TITLE
Proposal: Allow applying FieldEndianness to classes/structs

### ISSUE THREAD

**Issue by BADF00D** [jefffhaynes/BinarySerializer#96]
I just reviewer a code of my colleague, that uses BinarySerializer. I noticed, that you have to apply the FieldEndiannessAttribute to each field individually, rather then applying it to the class or struct. I think it's very uncommon, that the endianess switches within a structure. So what do you think, if you can apply the attribute at class level?

So rather then writing this:
```csharp
public class SomeData {
    [FieldEndianness(Endianness.Little)]
    [FieldOrder(0)]
    public ushort Id;

    [FieldEndianness(Endianness.Little)]
    [FieldOrder(1)]
    public short Order;
}
```
I want to write this:
```csharp
[FieldEndianness(Endianness.Little)]
public class SomeData {
    [FieldOrder(0)]
    public ushort Id;

    [FieldOrder(1)]
    public short Order;
}
```

In rare cases, where endianess swaps, the attribute nearest to the field, wins:
```csharp
[FieldEndianness(Endianness.Little)]
public class SomeData {
    [FieldOrder(0)]
    public ushort Id;

    [FieldOrder(1)]
    [FieldEndianness(Endianness.Big)]
    public short Order;
}
```

**Comment by jefffhaynes** [comment]
Have you tried it?  I would think that it should work. If not, it’s certainly a bug!

—
If you want to build a ship, don't drum up people to collect wood and don't assign them tasks and work, but rather teach them to long for the endless immensity of the sea.

Antoine de Saint-Exupery
________________________________

**Comment by BADF00D** [comment]
I tried an older version, but before I wrote this issue, I looked up the current implementation:

```csharp
/// <summary>
    ///     Specifies the endianness (value byte order) for a field.
    /// </summary>
    [AttributeUsage(AttributeTargets.Field | AttributeTargets.Property, AllowMultiple = true)]
    public sealed class FieldEndiannessAttribute : FieldBindingBaseAttribute, IConstAttribute
```

So the attribute can't be applied to classes and structs, because AttributeTargets.Class and AttributeTargets.Struct is missing. 
Or did I miss something?

**Comment by jefffhaynes** [comment]
No, that’s weird. I’ll take a look, thanks.

—
If you want to build a ship, don't drum up people to collect wood and don't assign them tasks and work, but rather teach them to long for the endless immensity of the sea.

Antoine de Saint-Exupery
________________________________

**Comment by jefffhaynes** [comment]
Sorry, I was reading this on my phone and I completely misunderstood what you were asking.  Currently nothing in binary serializer is applied in terms of type attributes and I'm not totally sure that would be an advantage.  Can your colleague apply the attribute to the appropriate field in the containing parent object and would that not accomplish the same thing?  Or alternatively, endianness can be passed in during serialization/deserialization.

The reason I like this approach is it allows you to reuse types with different attributes without having to redeclare them.  As a concrete example, the ISO9660 spec has structures appearing over and over with alternating endiannesses.

Anyway, I'm not saying one way is right or wrong but adding support for type-level declarations would be no small thing.  I'll think about it some more but I'd be curious to hear more about your specific use case and why you think it would be the better approach.

Thanks

**Comment by BADF00D** [comment]
We read data send from medical devices via RS232/USB/LAN. Usually the manufacturers of the device defines the endianess once per protocol. In rare cases this is switching during protocol extensions. So we can use your approach and apply this setting during deserialization. But if we use this approach, a type would no longer describe itself, because the endianess is declared somewhere else.

Our goals can be summarized as:
- A type should describe itself.
- We want to write as less code as possible.

My colleague mentioned, that the endianess attribute distracts the reader from the information that is really interesting at this point: the field offset. 

Imaging a class with 10 fields. 
```csharp
public class SomeData {
    [FieldEndianness(Endianness.Little)]
    [FieldOrder(0)]
    public ushort Field0;

    [FieldEndianness(Endianness.Little)]
    [FieldOrder(1)]
    public ushort Field1;

    [FieldEndianness(Endianness.Little)]
    [FieldOrder(2)]
    public ushort Field2;

    [FieldEndianness(Endianness.Little)]
    [FieldOrder(3)]
    public ushort Field3

    [FieldEndianness(Endianness.Little)]
    [FieldOrder(4)]
    public ushort Field4;

    [FieldEndianness(Endianness.Little)]
    [FieldOrder(5)]
    public ushort Field5;

    [FieldEndianness(Endianness.Little)]
    [FieldOrder(6)]
    public ushort Field6;

    [FieldEndianness(Endianness.Little)]
    [FieldOrder(7)]
    public ushort Field7;

    [FieldEndianness(Endianness.Little)]
    [FieldOrder(8)]
    public ushort Field8;

    [FieldEndianness(Endianness.Little)]
    [FieldOrder(9)]
    public ushort Field9;
}
```
In our case, the endianess does not change so the code could be simplified.
```csharp
[FieldEndianness(Endianness.Little)]
public class SomeData {

    [FieldOrder(0)]
    public ushort Field0;

    [FieldOrder(1)]
    public ushort Field1;

    [FieldOrder(2)]
    public ushort Field2;

    [FieldOrder(3)]
    public ushort Field3

    [FieldOrder(4)]
    public ushort Field4;

    [FieldOrder(5)]
    public ushort Field5;

    [FieldOrder(6)]
    public ushort Field6;

    [FieldOrder(7)]
    public ushort Field7;

    [FieldOrder(8)]
    public ushort Field8;

    [FieldOrder(9)]
    public ushort Field9;
}
```

But in the end, its a convenience feature. I have no hard facts, for or again this approach, it was just an idea. Besides the work needed to implement this, can you imaging any disadvantage? Do you think it can harm the readability in some circumstances?

**Comment by jefffhaynes** [comment]
Right, I definitely see what you’re saying but why not simply add a parent class like this

‘’’c#
public class Container
{
   [FieldEndianess(Endianess.Big)]
   public SomeData Data { get; set; }
}
‘’’

Doesn’t that accomplish the same thing?

**Comment by BADF00D** [comment]
Your approach would work, but ``SomeClass`` would not be self describing any more. 
With the current implementation, we prefer declaring declare the attribute at the field.

What do you mean with "raises issues of conflicting attribute resolution"? Do you think it might not be clear, which Attribute (at class or field level) has precedence over each other? I think it could be something like: the nearer the attribute it to the definition, the more precedence it takes.

E.g.:
```csharp
public class Container
{
   [FieldEndianness(Endianness.Big)] //priority 1
   public SomeData Data { get; set; }
}
[FieldEndianness(Endianness.Little)] //priority 2
public class SomeData {

    [FieldOrder(0)]
    public ushort Field0;

    [FieldOrder(1)]
    [FieldEndianness(Endianness.Big)] //prority 3 - highest
    public ushort Field1;
...
}
```

**Comment by jefffhaynes** [comment]
I think the type definition would have to be the default configuration and anything specified in the object graph would be considered overriding.

Anyway, I’ll play around with it and see what it would take.

I’m still not completely convinced that tightly coupling the type definition to what would essentially be the architecture (endianness) context makes sense but maybe I’m thinking about it wrong.

—
If you want to build a ship, don't drum up people to collect wood and don't assign them tasks and work, but rather teach them to long for the endless immensity of the sea.

Antoine de Saint-Exupery
________________________________

**Comment by BADF00D** [comment]
Thanks for thinking about it.

Btw: Great work!

**Comment by jefffhaynes** [comment]
Thank you, I’m glad it’s useful!

—
If you want to build a ship, don't drum up people to collect wood and don't assign them tasks and work, but rather teach them to long for the endless immensity of the sea.

Antoine de Saint-Exupery
________________________________

**Comment by danielmue** [comment]
@jefffhaynes If I haven't said it yet: Thank you for your work! Your library is really helpful.

**Comment by jefffhaynes** [comment]
Great, thank you!

—
If you want to build a ship, don't drum up people to collect wood and don't assign them tasks and work, but rather teach them to long for the endless immensity of the sea.

Antoine de Saint-Exupery
________________________________
